core/src/main/java/org/apache/carbondata/core/datamap/DistributableDataMapFormat.java

```diff
@@ -85,33 +143,67 @@
+        distributable.getDistributable().getSegment().setCacheable(!isFallbackJob);
+        distributable.getDistributable().getSegment().setReadCommittedScope(readCommittedScope);
+        List<Segment> segmentsToLoad = new ArrayList<>();
+        segmentsToLoad.add(distributable.getDistributable().getSegment());
         if (isJobToClearDataMaps) {
-          // if job is to clear datamaps just clear datamaps from cache and return
-          DataMapStoreManager.getInstance()
-              .clearDataMaps(table.getCarbonTableIdentifier().getTableUniqueName());
-          // clear the segment properties cache from executor
-          SegmentPropertiesAndSchemaHolder.getInstance()
-              .invalidate(table.getAbsoluteTableIdentifier());
-          blockletIterator = Collections.emptyIterator();
+          if (StringUtils.isNotEmpty(dataMapToClear)) {
+            List<TableDataMap> dataMaps =
+                DataMapStoreManager.getInstance().getAllVisibleDataMap(table);
+            List<TableDataMap> dataMapsTobeRemoved = new ArrayList<>();
+            for (TableDataMap tableDataMap : dataMaps) {
+              if (tableDataMap != null && dataMapToClear
+                  .equalsIgnoreCase(tableDataMap.getDataMapSchema().getDataMapName())) {
+                tableDataMap.deleteSegmentDatamapData(
+                    ((DataMapDistributableWrapper) inputSplit).getDistributable().getSegment()
+                        .getSegmentNo());
+                dataMapsTobeRemoved.add(tableDataMap);
+              }
+            }
+            dataMaps.removeAll(dataMapsTobeRemoved);
+            DataMapStoreManager.getInstance().getAllDataMaps().put(table.getTableUniqueName(),
+                dataMaps);
+          } else {
+            // if job is to clear datamaps just clear datamaps from cache and return
+            DataMapStoreManager.getInstance()
+                .clearDataMaps(table.getCarbonTableIdentifier().getTableUniqueName());
+            // clear the segment properties cache from executor
+            SegmentPropertiesAndSchemaHolder.getInstance()
+                .invalidate(table.getAbsoluteTableIdentifier());
+          }
+          List<ExtendedBlocklet> list = new ArrayList<ExtendedBlocklet>();
+          list.add(new ExtendedBlocklet());
+          blockletIterator = list.iterator();
           return;
+        } else if (invalidSegments.size() > 0) {
+          // clear the segmentMap and from cache in executor when there are invalid segments
+          DataMapStoreManager.getInstance().clearInvalidSegments(table, invalidSegments);
         }
-        dataMaps = tableDataMap.getTableDataMaps(distributable.getDistributable());
-        List<ExtendedBlocklet> blocklets = tableDataMap
-            .prune(dataMaps,
-                distributable.getDistributable(),
+        SegmentUpdateStatusManager updateStatusManager = new SegmentUpdateStatusManager(table);
```

Comment:
It is very heavy to read update status file and tablestatus file inside this class. Better create SegmentUpdateStatusManager for each segment in the driver and serialize and send here